--- a/kernel/runtime/ksud_integration.c
+++ b/kernel/runtime/ksud_integration.c
@@ -49,6 +49,19 @@
     "\n";
 // clang-format on
 
+char pavolia_rc_buf[4096] = {0};
+size_t pavolia_rc_len = 0;
+ssize_t pavolia_rc_pos = 0;
+
+void ksu_pavolia_add_prop(const char *prop, const char *val) {
+    char buf[256];
+    /* Using sys.boot_completed to inject when OS is ready */
+    snprintf(buf, sizeof(buf), "\non property:sys.boot_completed=1\n    setprop %s \"%s\"\n", prop, val);
+    strlcat(pavolia_rc_buf, buf, sizeof(pavolia_rc_buf));
+    pavolia_rc_len = strlen(pavolia_rc_buf);
+}
+EXPORT_SYMBOL(ksu_pavolia_add_prop);
+
 static void stop_init_rc_hook();
 static void stop_execve_hook();
 
@@ -288,6 +301,8 @@
         goto append_ksu_rc;
     if (ksu_rc_pos >= ksu_rc_len && module_rc_pos < module_rc_len)
         goto append_module_rc;
+    if (ksu_rc_pos >= ksu_rc_len && module_rc_pos >= module_rc_len && pavolia_rc_pos < pavolia_rc_len)
+        goto append_pavolia_rc;
 
     ret = orig_read(file, buf, count, pos);
     if (ret != 0) {
@@ -296,7 +311,7 @@
-    if (ksu_rc_pos >= ksu_rc_len && module_rc_pos >= module_rc_len) {
+    if (ksu_rc_pos >= ksu_rc_len && module_rc_pos >= module_rc_len && pavolia_rc_pos >= pavolia_rc_len) {
         return ret;
     }
     pr_info("read_proxy: orig read finished, start append rc\n");
 
@@ -334,6 +349,23 @@
         }
     }
 
+append_pavolia_rc:
+    if (pavolia_rc_pos < pavolia_rc_len && (size_t)ret < count) {
+        append_count = pavolia_rc_len - pavolia_rc_pos;
+        if (append_count > count - ret)
+            append_count = count - ret;
+        if (copy_to_user(buf + ret, pavolia_rc_buf + pavolia_rc_pos, append_count)) {
+            pr_info("read_proxy: pavolia append error, totally appended %zd\n", pavolia_rc_pos);
+            return ret;
+        }
+        pr_info("read_proxy: append pavolia %zu\n", append_count);
+        pavolia_rc_pos += append_count;
+        ret += append_count;
+        if (pavolia_rc_pos == (ssize_t)pavolia_rc_len) {
+            pr_info("read_proxy: pavolia append done\n");
+        }
+    }
+
     return ret;
 }
 
@@ -345,6 +377,8 @@
         goto append_ksu_rc;
     if (ksu_rc_pos >= ksu_rc_len && module_rc_pos < module_rc_len)
         goto append_module_rc;
+    if (ksu_rc_pos >= ksu_rc_len && module_rc_pos >= module_rc_len && pavolia_rc_pos < pavolia_rc_len)
+        goto append_pavolia_rc;
 
     ret = orig_read_iter(iocb, to);
     if (ret != 0) {
@@ -353,7 +387,7 @@
-    if (ksu_rc_pos >= ksu_rc_len && module_rc_pos >= module_rc_len) {
+    if (ksu_rc_pos >= ksu_rc_len && module_rc_pos >= module_rc_len && pavolia_rc_pos >= pavolia_rc_len) {
         return ret;
     }
     pr_info("read_iter_proxy: orig read finished, start append rc\n");
 
@@ -386,6 +420,21 @@
         }
     }
+
+append_pavolia_rc:
+    if (pavolia_rc_pos < pavolia_rc_len) {
+        append_count = copy_to_iter(pavolia_rc_buf + pavolia_rc_pos, pavolia_rc_len - pavolia_rc_pos, to);
+        if (!append_count) {
+            pr_info("read_iter_proxy: pavolia append error, appended %zd\n", pavolia_rc_pos);
+            return ret;
+        }
+        pr_info("read_iter_proxy: append pavolia %zu\n", append_count);
+        pavolia_rc_pos += append_count;
+        ret += append_count;
+        if (pavolia_rc_pos == (ssize_t)pavolia_rc_len) {
+            pr_info("read_iter_proxy: pavolia append done\n");
+        }
+    }
     return ret;
 }
 
@@ -581,10 +630,10 @@
     if (is_rc) {
         void __user *st_size_ptr = statbuf + offsetof(struct stat, st_size);
         long size, new_size;
-        size_t extra = ksu_rc_len + module_rc_len;
+        size_t extra = ksu_rc_len + module_rc_len + pavolia_rc_len;
         if (!copy_from_user_nofault(&size, st_size_ptr, sizeof(long))) {
             new_size = size + extra;
-            pr_info("adding rc len: %ld -> %ld (static=%zu module=%zu)", size, new_size, ksu_rc_len, module_rc_len);
+            pr_info("adding rc len: %ld -> %ld (static=%zu module=%zu pavolia=%zu)", size, new_size, ksu_rc_len, module_rc_len, pavolia_rc_len);
             if (!copy_to_user_nofault(st_size_ptr, &new_size, sizeof(long))) {
                 pr_info("added rc len");
             } else {
